### PR TITLE
Ports Bay JSON saves.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
Compiled port of:
https://github.com/Baystation12/Baystation12/pull/29635
https://github.com/Baystation12/Baystation12/pull/29659
https://github.com/Baystation12/Baystation12/pull/29665

Thanks @mloc!

Tested locally, seems to all work fine. Loaded my testing saves from legacy format and displayed them all properly, including robolimbs and markings.

### CODE REVIEW

File: code/modules/client/preferences.dm [hunk 1]
```diff
@@ -72,20 +76,71 @@ var/global/list/time_prefs_fixed = list()
...
+	if(S["version"] != 17)
+		return 0
+
+	// Legacy version 17 ~= new version 1
+	var/datum/pref_record_reader/savefile/savefile_reader = new(S, 1)
+
+	player_setup.load_preferences(savefile_reader)
+	var/orig_slot = default_slot
+
+	S.cd = "/torch"
```
Comment (quardbreak) [inline code/modules/client/preferences.dm]: `S.cd = "/[global.using_map.path]`

File: code/modules/client/preferences.dm [hunk 2]
```diff
@@ -72,20 +76,71 @@ var/global/list/time_prefs_fixed = list()
...
+	for(var/slot = 1 to 40)
+		if(!S.dir.Find("character[slot]"))
+			continue
+		S.cd = "/torch/character[slot]"
+		default_slot = slot
+		player_setup.load_character(savefile_reader)
+		save_character(override_key="character_torch_[slot]")
```
Comment (quardbreak) [inline code/modules/client/preferences.dm]: `save_character(override_key="character_[global.using_map.path]_[slot]")`

File: code/modules/client/preferences.dm [hunk 3]
```diff
@@ -72,20 +76,71 @@ var/global/list/time_prefs_fixed = list()
...
+		S.cd = "/torch"
```
Comment (quardbreak) [inline code/modules/client/preferences.dm]: `S.cd = "/[global.using_map.path]"`

File: code/modules/client/preferences.dm [hunk 4]
```diff
@@ -72,20 +76,71 @@ var/global/list/time_prefs_fixed = list()
...
+	var/datum/pref_record_reader/savefile/savefile_reader = new(S, 1)
+
+	player_setup.load_preferences(savefile_reader)
+	var/orig_slot = default_slot
+
+	S.cd = "/torch"
+	for(var/slot = 1 to 40)
+		if(!S.dir.Find("character[slot]"))
+			continue
+		S.cd = "/torch/character[slot]"
```
Comment (quardbreak) [inline code/modules/client/preferences.dm]: `S.cd = "/[global.using_map.path]/character[slot]"`

**Review by quardbreak — CHANGES_REQUESTED** [review]

File: code/modules/client/preferences.dm (hunk 1 above)
Comment (MistakeNot4892) [reply code/modules/client/preferences.dm]: Yep, sorted this out earlier, just hadn't pushed.

**Comment by MistakeNot4892** [comment]
Tested locally, seems to all work fine. Loaded my testing saves from legacy format and displayed them all properly, including robolimbs and markings.

**Review by afterthought2 — APPROVED** [review]

**Review by comma — APPROVED** [review]

**Comment by Lohikar** [comment]
Let's break things.

**Pull request merged**